src/iop/hazeremoval.c

```diff
@@ -0,0 +1,962 @@
+                 img2_bak.data,
+                 img2.data + i0, img1.width,
+                 w);
+    }
+    free_gray_image(&img2_bak);
+  }
+}
+
+// calculate the dark channel (minimal color component over a box of size (2*w+1) x (2*w+1) )
+static void dark_channel(rgb_image img1, gray_image img2, int w)
+{
+  size_t size = (size_t)img1.height * (size_t)img1.width;
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static)
+#endif
+  for (size_t i = 0; i < size; i++)
+  {
+    float *pixel = get_rgb_pixel(img1, i);
+    float m = pixel[0];
+    m = min_f(pixel[1], m);
+    m = min_f(pixel[2], m);
+    img2.data[i] = m;
+  }
+  box_min(img2, img2, w);
+}
+
+// calculate the transition map
+static void transition_map(rgb_image img1, gray_image img2, int w, rgb_pixel A0, float strength)
+{
+  size_t size = (size_t)img1.height * (size_t)img1.width;
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static)
+#endif
+  for (size_t i = 0; i < size; i++)
+  {
+    float *pixel = get_rgb_pixel(img1, i);
+    float m = pixel[0] / A0[0];
+    m = min_f(pixel[1] / A0[1], m);
+    m = min_f(pixel[2] / A0[2], m);
+    img2.data[i] = 1.f - m * strength;
+  }
+  box_max(img2, img2, w);
+}
+
+// apply guided filter to single-component image img using the 3-components
+// image imgg as a guide
+static void guided_filter_tiling(rgb_image imgg, gray_image img, tile target, int w, float eps)
+{
+  const tile source = { max_i(target.left - w, 0),
+			min_i(target.right + w , imgg.width),
+			max_i(target.lower - w , 0),
+			min_i(target.upper + w , imgg.height) };
+  const int width = source.right - source.left;
+  const int height = source.upper - source.lower;
+  size_t size = (size_t)width * (size_t)height;
+  gray_image imgg_mean_r = new_gray_image(width, height);
+  gray_image imgg_mean_g = new_gray_image(width, height);
+  gray_image imgg_mean_b = new_gray_image(width, height);
+  gray_image img_mean = new_gray_image(width, height);
+  for (int j_imgg = source.lower; j_imgg < source.upper; j_imgg++)
+  {
+    int j = j_imgg - source.lower;
+    for (int i_imgg = source.left; i_imgg < source.right; i_imgg++)
+    {
+      int i = i_imgg - source.left;
+      float *pixel = get_rgb_pixel(imgg, i_imgg + (size_t)j_imgg * imgg.width);
+      size_t k = i + (size_t)j * width;
+      imgg_mean_r.data[k] = pixel[0];
+      imgg_mean_g.data[k] = pixel[1];
+      imgg_mean_b.data[k] = pixel[2];
+      img_mean.data[k] = img.data[i_imgg + (size_t)j_imgg * img.width];
+    }
+  }
+  box_mean(imgg_mean_r, imgg_mean_r, w);
+  box_mean(imgg_mean_g, imgg_mean_g, w);
+  box_mean(imgg_mean_b, imgg_mean_b, w);
+  box_mean(img_mean, img_mean, w);
+  gray_image cov_imgg_img_r = new_gray_image(width, height);
+  gray_image cov_imgg_img_g = new_gray_image(width, height);
+  gray_image cov_imgg_img_b = new_gray_image(width, height);
+  gray_image var_imgg_rr, var_imgg_rg, var_imgg_rb,
+    var_imgg_gg, var_imgg_gb,
+    var_imgg_bb;
+  var_imgg_rr = new_gray_image(width, height);
```

Comment:
these 6 `var_imgg_*` variables only differ by 1-2 last symbols.
this is the best place to hide a problem that may never be found.
I.e. it's a horrible mess, even though i understand why it is the way it is.